src/lib/storage/value_column.cpp

```diff
@@ -155,6 +155,36 @@ std::shared_ptr<BaseColumn> ValueColumn<T>::copy_using_allocator(const Polymorph
   }
 }
 
+template <typename T>
+MemoryUsage ValueColumn<T>::estimate_memory_usage(MemoryUsageEstimationMode estimation_mode) const {
+  return MemoryUsage{_values.size() * sizeof(T) + (_null_values ? _null_values->size() * sizeof(bool) : 0u)};
+}
+
+template <>
+MemoryUsage ValueColumn<std::string>::estimate_memory_usage(MemoryUsageEstimationMode estimation_mode) const {
+  auto bytes = size_t{(_null_values ? _null_values->size() * sizeof(bool) : 0u) + sizeof(*this)};
+
+  switch (estimation_mode) {
+    case MemoryUsageEstimationMode::Fast:
+      return MemoryUsage{bytes + _values.size() * sizeof(std::string)};
+    case MemoryUsageEstimationMode::MoreExact: {
+      for (const auto& string : _values) {
+        if (string.size() < sizeof(std::string)) {
+          // Assume small string optimization is used
+          // https://blogs.msmvps.com/gdicanio/2016/11/17/the-small-string-optimization/
+          bytes += sizeof(std::string);
+        } else {
+          bytes += sizeof(std::string) + string.size();
+        }
+      }
+      return MemoryUsage{bytes};
+    }
+  }
+
+  Fail("Clang thinks this might be reachable, can't see how it would be...");
```

Comment:
Adding the default case to the switch statement might help.

Reply:
would do the same thing as this does, except that the switch statement (which includes all enum cases, for god's sake, clang!) would become dirty.

Reply:
Just move the fail into the switch statement as the default case.

Reply:
Minor discussion, but no. The switch statement is totally fine by any standard, adding a default case just to make a particular compiler shut up lowers the quality of the code construct, because the mere presence of the default case, even if the string in the Fail explains it, suggests to the reader that there are unhandled cases.

Reply:
The reason why clang thinks the lines below could be reach is that with C++17, you can write things like: `MemoryUsageEstimationMode{14}`.

Reply:
What the smoking fuck? Doesn't that destroy most of the advantage of enum class? Out of interest, do you have any source as to why they allow that now?

Reply:
So there is no telling from the absence of the default arg in combination with the code compiling that all cases were enumerated anymore... you failed, c++ committee, you failed.

Reply:
Yes, one use case is to use empty enums as IDs. 

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0138r2.pdf